cmd/maya-apiserver/spc-watcher/lease.go

```diff
@@ -86,18 +177,79 @@ func (sl *spcLease) UpdateLease(leaseValue string) (*apis.StoragePoolClaim, erro
 	return spcObject, nil
 }
 
-// Can be used for reconcile loop use cases
-// TODO Remove using patch instead of update
-func (sl *spcLease) RemoveLease() *apis.StoragePoolClaim {
-	spcObject, err := sl.UpdateLease("")
+func (sl *spcLease) RemoveLease() {
+	_, err := sl.patchSpc()
 	if err != nil {
-		runtime.HandleError(fmt.Errorf("Lease could not be removed:%v", err))
+		newErr := fmt.Errorf("Lease could not be removed:%v", err)
+		runtime.HandleError(newErr)
 	}
-	return spcObject
+	glog.Info("Lease removed successfully on storagepoolclaim")
 }
 
 func (sl *spcLease) getPodName() string {
-	podName := menv.Get(menv.OpenEBSMayaPodName)
-	podNameSpace := menv.Get(menv.OpenEBSNamespace)
+	podName := env.Get(env.OpenEBSMayaPodName)
+	podNameSpace := env.Get(env.OpenEBSNamespace)
 	return podNameSpace + "/" + podName
 }
+
+// patchSpc will patch the spc object to release the lease
+func (sl *spcLease) patchSpc() (*apis.StoragePoolClaim, error) {
+	spcPatch := make([]SPCPatch, 1)
+	// setting operation as remove
+	spcPatch[0].Op = PatchOperation
+	// object to be removed is finalizers
+	spcPatch[0].Path = PatchPath
+	leaseValueObj, err := parseLeaseValue(sl.spcObject.Annotations[SpcLeaseKey])
+	leaseValueObj.HolderIdentity = ""
+	newLeaseValue, err := json.Marshal(leaseValueObj)
+	if err != nil {
+		return nil, err
+	}
+	spcPatch[0].Value = string(newLeaseValue)
+	spcPatchJSON, err := json.Marshal(spcPatch)
+	if err != nil {
+		glog.Errorf("Error marshalling spcPatch object: %s", err)
+	}
+	obj, err := sl.oecs.OpenebsV1alpha1().StoragePoolClaims().Patch(sl.spcObject.Name, types.JSONPatchType, spcPatchJSON)
+	return obj, err
+}
+
+// checkLeaderLiveness checks whether the holder of lease is live or not
+// If the holder is not live or does not exists the function will return true.
+
+// If the holder of lease is not live or does not exists the lease can be acquired
+// by the other contestant(i.e. maya pod)
+func (sl *spcLease) checkLeaderLiveness(leaseValueObj lease) bool {
```

Comment:
Try to think/design in terms of entities that come up during the course of writing the logic.
e.g. Leader, Lease, Patcher comes up as entities that can be considered as first class structs/interfaces.

Reply:
Again Leader and Patcher are tight bound to SPC and it is required as those function need to know the object type.
Also, if use cases come we can improve. Right now can't think of making it loosely coupled due to a constraint i.e. object type should be known or methods to manipulate on objects that will under go lease algo.